spark/src/main/scala/org/apache/spark/sql/delta/commands/alterDeltaTableCommands.scala

```diff
@@ -586,16 +586,31 @@ case class AlterTableChangeColumnDeltaCommand(
     recordDeltaOperation(deltaLog, "delta.ddl.alter.changeColumns") {
       val txn = startTransaction()
       val metadata = txn.metadata
-      val oldSchema = metadata.schema
+      val oldSchema = if (
+        sparkSession.conf.get(DeltaSQLConf.DELTA_BYPASS_CHARVARCHAR_TO_STRING_FIX)) {
+          metadata.schema
+        } else {
+          SchemaUtils.getRawSchemaWithoutCharVarcharMetadata(metadata.schema)
+        }
       val resolver = sparkSession.sessionState.conf.resolver
 
       // Verify that the columnName provided actually exists in the schema
       SchemaUtils.findColumnPosition(columnPath :+ columnName, oldSchema, resolver)
 
-      val newSchema = transformSchema(oldSchema, Some(columnName)) {
+      var newSchema = transformSchema(oldSchema, Some(columnName)) {
```

Comment:
nit: maybe make it a `val transformedSchema = ...` here
and
```
    val newSchema =  if (!sparkSession.conf.get(DeltaSQLConf.DELTA_BYPASS_CHARVARCHAR_TO_STRING_FIX)) {
        // Fields with type CHAR/VARCHAR had been converted from
        // (StringType, metadata = 'VARCHAR(n)') into (VARCHAR(n), metadata = '')
        // so that CHAR/VARCHAR to String conversion can be handled correctly. We should
        // invert this conversion so that downstream operations are not affected.
        newSchema = CharVarcharUtils.replaceCharVarcharWithStringInSchema(transformedSchema)
      } else {
        transformedSchema
      }
```
below

Reply:
Done. Also, updated the code so that we invoke sparkSession.conf.get(DeltaSQLConf.DELTA_BYPASS_CHARVARCHAR_TO_STRING_FIX) once and store it in a val which is reused:
https://github.com/delta-io/delta/pull/3346/files#diff-3f4163d260854433e36f91a05268df8404b5479a2de432044e732f2563a9e66eR589